core/dynamo.c

```diff
@@ -978,7 +984,22 @@ dynamo_shared_exit(thread_record_t *toexit /* must ==cur thread for Linux */
      * client trying to use api routines that depend on fragment state.
      */
     instrument_exit();
-#endif
+# ifdef CLIENT_SIDELINE
+    /* We only need do a second synch-all if there are sideline client threads. */
```

Comment:
The condition was the conditional macro, but I guess add a check is not too bad.

If the client thread create another thread, dynamo_thread_init will be called, and it will wait for app exit.
We split the synch-all is to let sideline threads have a chance to exit gracefully. If it creates another thread, which will be terminated when process terminates, which seems fine for me.
It could do many more crazy things, e.g., fork and exec a different program. I am not sure if we can/want to handle all of it.